### PR TITLE
Stop explicitly manipulating firewalls in `pbench-fio` and `pbench-uperf`

### LINKED ISSUE #1217: Stop using `firewall-cmd` in `pbench-fio`
Today, `pbench-fio` changes a remote system as a side-effect of executing.

Why not require the caller to open the port in their environment before running `pbench-fio`?

And we don't restore the port state any where, it seems, so it might be best to leave this to the caller of `pbench-fio` to make sure the proper port is opened.

Arguably we should document this behavioral need as well.

### PR DESCRIPTION
**Pull request by portante**
All three existing benchmark scripts which manipulate network firewalls have been changed to ignore firewalls entirely.

For `pbench-fio`, this means it no longer attempts to add open ports when clients are used.

For `pbench-uperf` and `pbench-netperf`, those scripts no longer disable `firewalld` explicitly.

Any one using `pbench-fio` in client mode, or either `pbench-uperf` or `pbench-netperf`, must be responsible for properly enabling access on the expected ports.

The documentation for `pbench-uperf` and `pbench-fio` have been updated to reflect this new behavior.

Fixed #1217.

### CODE REVIEW

File: agent/bench-scripts/test-bin/bm
Comment (dbutenhof) [inline agent/bench-scripts/test-bin/bm:13]: Indentation?

File: agent/bench-scripts/pbench-fio.md
Comment (dbutenhof) [inline agent/bench-scripts/pbench-fio.md:353]: Ha!

**Review by dbutenhof — APPROVED** [review]
I think the indentation is funky in one place; but otherwise just a few side comments. Although given the assorted fixes across the layers of commit here, I do wonder about whether you should consider changing the PR title...

File: agent/bench-scripts/pbench-fio
Comment (webbnh) [reply agent/bench-scripts/pbench-fio:317]: OK, this code idiom is annoyingly clever.  (I'm not sure whether I want to remember this for future use or forget I ever saw it!)

**Review by webbnh — CHANGES_REQUESTED** [review]
This could use some polishing.

File: agent/bench-scripts/pbench-fio
Comment (portante) [reply agent/bench-scripts/pbench-fio:317]: I'd rather keep the code the way it is since it does not need to go back and fix anything up, and it is the same number of lines, and touches the values less.

File: agent/bench-scripts/test-bin/bm
Comment (portante) [reply agent/bench-scripts/test-bin/bm:13]: Hmm, isn't `ip_name` used on line 13, and `client` used on line 14?

File: agent/bench-scripts/pbench-fio.md
Comment (portante) [reply agent/bench-scripts/pbench-fio.md:353]: This line was taken from `pbench-uperf.md`.

File: agent/bench-scripts/pbench-fio
Comment (portante) [reply agent/bench-scripts/pbench-fio:184]: Yes, the user asked for `pbench-fio --help`, that is not an error.

File: agent/bench-scripts/pbench-fio
Comment (portante) [reply agent/bench-scripts/pbench-fio:503]: Array reference.

File: agent/bench-scripts/pbench-fio
Comment (portante) [reply agent/bench-scripts/pbench-fio]: Too much output each sample of each iteration.  And not to reveal the end of the story, but all this goes away anyways.

File: agent/bench-scripts/pbench-fio
Comment (portante) [reply agent/bench-scripts/pbench-fio]: The double quote before `pushd` makes it all work.

File: agent/bench-scripts/pbench-fio
Comment (portante) [reply agent/bench-scripts/pbench-fio:676]: Very simply, if you don't enclose those variable expansions in quotes, then you end up passing potentially more than 2 arguments.  The quotes ensure we pass 2 arguments.

As for quoting internally, from past experience debugging when these shell scripts fail due to variable expansion issues, it is easier to track down when you can match argument counts provided to those expected.  Then when an error occurs, we know where to look.

In this case no matter what we do, we'll have a problem.  With them enclosed in double quotes, the `$0` and `$1` inner references will expand with the proper values, but the `bash -c` script will fail entirely, where it might not if the resulting variables in the `bash -c` script happen to work.

File: agent/bench-scripts/pbench-fio
Comment (portante) [reply agent/bench-scripts/pbench-fio:444]: > So, by default we're going to _skip_ the pre-check?

Yes.  The use of `pbench-fio --pre-check` is only when we have remote clients and we need to run the same check we run locally on the remote clients.

>  Is that "good"?...I guess that's an extension of the old behavior, right?

Yes, you'll notice that `--install` was not in the `usage` text, and neither is `--pre-check`.  Basically it is an internal-only parameter so that we can share the same basic sanity checks locally and remotely.

File: agent/bench-scripts/pbench-fio
Comment (portante) [reply agent/bench-scripts/pbench-fio:576]: Once we discover that a specified block device does not exist, we want to fail the local pre-check.

File: agent/bench-scripts/tests/pbench-fio/test-32.txt
Comment (portante) [reply agent/bench-scripts/tests/pbench-fio/test-32.txt:118]: Bugger all, that is a bug.

File: agent/bench-scripts/pbench-fio
Comment (portante) [reply agent/bench-scripts/pbench-fio:671]: And the more I look at the double-single quote thing the more I don't understand what it is doing.

File: agent/bench-scripts/pbench-fio
Comment (portante) [reply agent/bench-scripts/pbench-fio]: After reviewing this further, I am not sure I understand out the quoting is working.

File: agent/bench-scripts/pbench-fio
Comment (portante) [reply agent/bench-scripts/pbench-fio]: Yeah, I leaked future changes back into the previous commits.

File: agent/bench-scripts/pbench-fio
Comment (portante) [reply agent/bench-scripts/pbench-fio]: I'll see how leaving one set of single quotes around the `bash -c` string works.

File: agent/bench-scripts/pbench-fio.md
Comment (portante) [reply agent/bench-scripts/pbench-fio.md:364]: Adjusted the text to include the use of `--unique-ports`, but not sure we want to change the port numbers used with this commit.

It is `fio` the has a default client port of `8765`, so that is pretty ingrained in the usage pattern today.

File: agent/bench-scripts/pbench-fio
Comment (portante) [reply agent/bench-scripts/pbench-fio]: And while tracking down the `clients` variable usage I got tired of two different reference patterns.

**Comment by portante** [comment]
FWIW, I created PR #2526 to track the `pbench-fio` usage changes in a separate commit along with some related cleanup.

**Comment by portante** [comment]
I've decided to convert this to a `Draft` so that I can break this PR out a bit.  The first break-out is PR #2526.  Once that is accepted and merged, I'll post another break-out PR.

File: agent/bench-scripts/pbench-fio
Comment (dbutenhof) [reply agent/bench-scripts/pbench-fio]: > And the more I look at the double-single quote thing the more I don't understand what it is doing.

Especially the empty `''` block. The second single-quote would seem to protect the variables from expansion here, deferring to the sub-bash under the screen (weird enough already); but I'm not sure I understand why since if they have any value at all in that context (i.e., they're exported ... are they?), there's no mechanism for the expansion to differ in that context. Bee-zarre.

File: agent/bench-scripts/pbench-fio
Comment (dbutenhof) [reply agent/bench-scripts/pbench-fio:197]: My initial inclination with your redirections had been to say we should just redirect the individual `usage` printf commands; but, yeah, this special case complicates that encapsulation. I'm not really sure that a convention of dumping usage strings to stderr would be harmful even in `--help`... but it's fine.

(I experimented a bit; and while most commands seem to follow the convention of "error usage to stderr and -h/--help to stdout, there are exceptions ... `netstat -h` for example goes to stderr. Funny the things you just never notice until something makes you look ...)

**Review by dbutenhof — APPROVED** [review]

File: agent/bench-scripts/pbench-fio
Comment (webbnh) [reply agent/bench-scripts/pbench-fio:315]: I think I disagree with most of those assertions, but I don't object to this proposed change.

Comment (webbnh) [reply agent/bench-scripts/pbench-fio:315]: (I'm leaving it up to @dbutenhof to resolve this conversation, though. 😉)

File: agent/bench-scripts/test-bin/bm
Comment (webbnh) [reply agent/bench-scripts/test-bin/bm]: Ah, yes.  (That's what I get for doing reviews too long into the day....)

File: agent/bench-scripts/pbench-fio.md
Comment (webbnh) [reply agent/bench-scripts/pbench-fio.md]: My real concern for this comment was the accuracy of the text wrt when it uses sequential ports (and, I presume you've addressed that).

I'm less concerned about using a specific port (like 8765) than I am about the sequential usage covering a substantial amount of the otherwise "reserved" port range.  But, if an issue with that actually arises, I presume the user can specify `--unique-ports` together with a starting port in the unreserved range and resolve the problem.

File: agent/bench-scripts/test-bin/bm
Comment (portante) [reply agent/bench-scripts/test-bin/bm]: I believe this was addressed in PR #2527.

File: agent/bench-scripts/pbench-fio
Comment (portante) [reply agent/bench-scripts/pbench-fio]: This work has been moved to PR #2528.

File: agent/bench-scripts/pbench-fio
Comment (portante) [reply agent/bench-scripts/pbench-fio]: This work has been moved to PR #2528.

File: agent/bench-scripts/pbench-fio
Comment (portante) [reply agent/bench-scripts/pbench-fio]: Fixed.

File: agent/bench-scripts/pbench-fio
Comment (portante) [reply agent/bench-scripts/pbench-fio]: This work has been moved to PR #2528.

File: agent/bench-scripts/tests/pbench-fio/test-32.txt
Comment (portante) [reply agent/bench-scripts/tests/pbench-fio/test-32.txt]: Fixed in PR #2528.

File: agent/bench-scripts/pbench-fio
Comment (portante) [reply agent/bench-scripts/pbench-fio]: This work has been moved to PR #2528.

**Review by dbutenhof — APPROVED** [review]

**Review by dbutenhof — APPROVED** [review]

**Review by webbnh — APPROVED** [review]
GTG

**Pull request merged**
